source/common/mongo/proxy.cc

```diff
@@ -241,5 +245,54 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
   return DecoderPtr{new DecoderImpl(callbacks)};
 }
 
+FaultConfigImpl::FaultConfigImpl(const Json::Object& fault_config)
+    : delay_percent_(
+          static_cast<uint32_t>(fault_config.getObject("fixed_delay")->getInteger("percent"))),
+      duration_ms_(
+          static_cast<uint64_t>(fault_config.getObject("fixed_delay")->getInteger("duration_ms"))) {
+}
+
+Optional<uint64_t> ProxyFilter::delayDuration() {
+  Optional<uint64_t> result;
+
+  if (!fault_config_)
+    return result;
+
+  if (!runtime_.snapshot().featureEnabled("mongo.fault.delay.percent",
+                                          fault_config_->delayPercent())) {
+    return result;
+  }
+
+  uint64_t duration = runtime_.snapshot().getInteger("mongo.fault.delay.duration_ms",
+                                                     fault_config_->delayDuration());
+
+  // Delay only if the duration is > 0ms.
+  if (duration > 0) {
+    result.value(duration);
+  }
+
+  return result;
+}
+
+void ProxyFilter::delayInjectionTimerCallback() {
```

Comment:
What you have now is not correct. Mongo can be multiplexing, so you can get 2 queries in a decode for example. If the 2nd query continues while there is an outstanding delay operation, badness will occur. You can potentially handle like you are doing, but you need to consider this case. (There are other problems with this PR BTW but will let @ccaraman continue with review).